### PR TITLE
feat: sdk async key exchange

### LINKED ISSUE #10315: [Bug]: Wrong origin on confirmation screens
### Describe the bug

The origin on confirmation screens is currently wrong and it seems to pick up the latest seen origin in the inapp browser. 

This had been fixed before in:
https://github.com/MetaMask/metamask-mobile/pull/9914

When using Deeplink communication layer via iOS SDK the user is not directed back automatically when a transaction is confirmed.

### Expected behavior

_No response_

### Screenshots/Recordings

https://github.com/user-attachments/assets/4cf414f7-da94-4ff2-9f03-d34c847b7a8e

https://github.com/user-attachments/assets/a8bcfc38-69d4-4b7e-899e-f0acd2205c8d

### Steps to reproduce

GIVEN a user is connected to an SDK dapp
WHEN the user navigates to a dapp via inapp browser
AND triggers an action from the SDK dapp
THEN the user should see in the pill the origin for the SDK Dapp

Actual result:
The pill displays the last seen origin in the inapp browser instead

### Error messages or log output

_No response_

### Version

7.27.0

### Build type

None

### Device

-

### Operating system

iOS, Android

### Additional context

_No response_

### Severity

_No response_

### PR DESCRIPTION
**Pull request by abretonc7s**
## **Description**

- Full async key exchange implementation for SDK with update of comm layer.
image
- Full documentation at https://www.notion.so/metamask-consensys/e3cb7482a492445fb0665334e53a1795?v=34d0d322d99f493aaf7b1ac58ed1c87b&p=2c95933da20e465fa0eeeed48adb9e68&pm=s

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-mobile/issues/10315

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by abretonc7s** [comment]
@SocketSecurity ignore npm/utf-8-validate@5.0.10

**Review by andreahaku — DISMISSED** [review]
LGTM

**Comment by christopherferreira9** [comment]
Full testing strategy and results [here](https://www.notion.so/metamask-consensys/Async-key-exchange-886ae1272a4c4b6d9658996abf3c935c).

File: app/core/RPCMethods/RPCMethodMiddleware.ts [hunk 1]
```diff
@@ -612,6 +613,7 @@ export const getRpcMethodMiddleware = ({
         const pageMeta = {
           meta: {
             url: url.current,
+            channelId,
```
Comment (tommasini) [inline app/core/RPCMethods/RPCMethodMiddleware.ts:616]: Why is the channelId only needed for the `personal_sign` and not the others type sign rpc method?

File: app/core/SDKConnect/Connection/Connection.ts [hunk 2]
```diff
@@ -273,8 +269,18 @@ export class Connection extends EventEmitter2 {
     );
   }
 
-  public connect({ withKeyExchange }: { withKeyExchange: boolean }) {
-    return connect({ instance: this, withKeyExchange });
+  public connect({
+    withKeyExchange,
+    authorized,
```
Comment (tommasini) [inline app/core/SDKConnect/Connection/Connection.ts:274]: Sorry, I'm having some trouble understanding the need of this authorized, can we have TSDoc at this method?

File: app/core/SDKConnect/Connection/ConnectionLifecycle/connect.test.ts [hunk 3]
```diff
@@ -23,45 +23,56 @@ describe('connect', () => {
   });
 
   it('should log the connect action with channel ID and key exchange flag', () => {
-    connect({ instance: mockConnection, withKeyExchange: true });
+    connect({
```
Comment (tommasini) [inline app/core/SDKConnect/Connection/ConnectionLifecycle/connect.test.ts:26]: All the test cases are covering only authorized false, is possible to have scenarios where authorized is true, if they make sense?

File: index.js [hunk 4]
```diff
@@ -27,6 +27,7 @@ import { isNotificationsFeatureEnabled } from './app/util/notifications';
 import { Performance } from './app/core/Performance';
 Performance.setupPerformanceObservers();
 
+LogBox.ignoreAllLogs();
```
Comment (tommasini) [inline index.js:30]: Was this intentional?

File: app/core/SDKConnect/Connection/ConnectionLifecycle/connect.test.ts (hunk 3 above)
Comment (abretonc7s) [reply app/core/SDKConnect/Connection/ConnectionLifecycle/connect.test.ts:26]: yes but will work on that on another PR. This is a big protocol rewrite

File: index.js (hunk 4 above)
Comment (abretonc7s) [reply index.js:30]: no.... removing

File: app/core/SDKConnect/Connection/Connection.ts (hunk 2 above)
Comment (abretonc7s) [reply app/core/SDKConnect/Connection/Connection.ts:274]: this is used in the protocol (separate package), not on the mobile code

**Review by tommasini — APPROVED** [review]

**Comment by christopherferreira9** [comment]
E2E are currently experiencing flaky results:
The first run had 4 tests failing: https://app.bitrise.io/build/b1fb0923-0428-4e22-815d-d4b9b7ef6989
The second run had only 2 tests failing: https://app.bitrise.io/build/0bab6409-fa2a-42b1-9922-ed93b7809406

No fixes were done in between the two runs

**Comment by christopherferreira9** [comment]
E2E passing here: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/c0c9ddbb-a272-4622-9b17-33da70728e3e

**Pull request merged**